File: heap_dataset/heap_excel.py
data_awal = [
    (5288,"pensil"),(5993,"pulpen"),(8689,"penghapus"),(8043,"buku"),(8699,"sampul"),
    (2156,"penggaris"),(4457,"kertas"),(8938,"cat"),(2618,"stabilo"),(9033,"mobil"),
    (9971,"motor"),(3874,"becak"),(5914,"sepeda"),(2398,"kereta"),(3725,"pesawat"),
    (5210,"perahu"),(7363,"kapal"),(7631,"rakit"),(4513,"kipas"),(5656,"charger"),
    (6453,"peci"),(8783,"sarung"),(8194,"sajadah"),(9783,"smartphone"),(3685,"jam"),
    (4490,"televisi"),(8294,"laptop"),(8563,"komputer"),(1070,"mouse"),(5408,"keyboard"),
    (8258,"tablet"),(9309,"jendela"),(1138,"kaca"),(2751,"pintu"),(3258,"kompor"),
    (6402,"lemari"),(7921,"kasur"),(9781,"ranjang"),(3818,"bantal"),(5204,"baju"),
    (6119,"kaos"),(1928,"celana"),(4207,"mukena"),(7255,"jilbab"),(5309,"pigura"),
    (2897,"antena"),(8028,"kulkas"),(1660,"dispenser"),(3248,"meja"),(5641,"kursi"),
    (7376,"kemoceng"),(3525,"sapu"),(4492,"gayung"),(7187,"sabun"),(1305,"sikat"),
    (6602,"shampo"),(8153,"botol"),(3561,"gelas"),(5082,"piring"),(7151,"panci"),
    (7524,"wajan"),(9178,"blender"),(9817,"galon"),(4304,"cobek"),(6820,"termos"),
    (9151,"kran"),(3482,"selang"),(3316,"karpet"),(5192,"tikar"),(7572,"keset"),
    (7660,"sepatu"),(9224,"kaos kaki"),(5083,"jaket"),(6362,"piama"),(6465,"piano"),
    (9888,"gitar"),(4159,"angklung"),(4969,"suling"),(5097,"toples"),(6271,"parfum"),
    (9250,"sisir"),(3409,"topi"),(4577,"gunting"),(6244,"pisau"),(8612,"kaleng"),
    (4650,"tisu"),(6799,"tas"),(9298,"ikat pinggang"),(4361,"korek api"),(4379,"kopi"),
    (6928,"gula"),(3195,"cabai"),(5741,"wortel"),(6852,"timun"),(8147,"apel"),
    (8902,"jeruk"),(8967,"tomat"),(1302,"pisang"),(2363,"pepaya"),(6861,"bawang"),
]

# ===== MIN HEAP =====
min_heap = []

def min_swap(i, j):
    min_heap[i], min_heap[j] = min_heap[j], min_heap[i]

def min_up(i):
    while i > 0 and min_heap[(i-1)//2][0] > min_heap[i][0]:
        min_swap(i, (i-1)//2); i = (i-1)//2

def min_down(i):
    n = len(min_heap)
    while True:
        s, l, r = i, 2*i+1, 2*i+2
        if l < n and min_heap[l][0] < min_heap[s][0]: s = l
        if r < n and min_heap[r][0] < min_heap[s][0]: s = r
        if s == i: break
        min_swap(i, s); i = s

def min_insert(id, nama):
    min_heap.append((id, nama)); min_up(len(min_heap)-1)

def min_delete():
    if not min_heap: print("Min-Heap kosong!"); return
    id = int(input("Masukkan ID yang ingin dihapus dari Min-Heap: "))
    idx = next((i for i, x in enumerate(min_heap) if x[0] == id), -1)
    if idx == -1: print("ID tidak ditemukan!"); return
    print(f"Data dihapus: {min_heap[idx][0]} - {min_heap[idx][1]}")
    min_heap[idx] = min_heap[-1]; min_heap.pop()
    if idx < len(min_heap): min_up(idx); min_down(idx)

def min_print():
    if not min_heap: print("Min-Heap kosong!"); return
    sorted_data = sorted(min_heap, key=lambda x: x[0])
    print("\n=== DATA MIN-HEAP (Ascending) ===")
    print(f"{'ID':<10} {'Nama'}")
    print("-" * 25)
    for id, nama in sorted_data: print(f"{id:<10} {nama}")

# ===== MAX HEAP =====
max_heap = []

def max_swap(i, j):
    max_heap[i], max_heap[j] = max_heap[j], max_heap[i]

def max_up(i):
    while i > 0 and max_heap[(i-1)//2][0] < max_heap[i][0]:
        max_swap(i, (i-1)//2); i = (i-1)//2

def max_down(i):
    n = len(max_heap)
    while True:
        s, l, r = i, 2*i+1, 2*i+2
        if l < n and max_heap[l][0] > max_heap[s][0]: s = l
        if r < n and max_heap[r][0] > max_heap[s][0]: s = r
        if s == i: break
        max_swap(i, s); i = s

def max_insert(id, nama):
    max_heap.append((id, nama)); max_up(len(max_heap)-1)

def max_delete():
    if not max_heap: print("Max-Heap kosong!"); return
    id = int(input("Masukkan ID yang ingin dihapus dari Max-Heap: "))
    idx = next((i for i, x in enumerate(max_heap) if x[0] == id), -1)
    if idx == -1: print("ID tidak ditemukan!"); return
    print(f"Data dihapus: {max_heap[idx][0]} - {max_heap[idx][1]}")
    max_heap[idx] = max_heap[-1]; max_heap.pop()
    if idx < len(max_heap): max_up(idx); max_down(idx)

def max_print():
    if not max_heap: print("Max-Heap kosong!"); return
    sorted_data = sorted(max_heap, key=lambda x: x[0], reverse=True)
    print("\n=== DATA MAX-HEAP (Descending) ===")
    print(f"{'ID':<10} {'Nama'}")
    print("-" * 25)
    for id, nama in sorted_data: print(f"{id:<10} {nama}")

# ===== INIT & MAIN =====
for id, nama in data_awal:
    min_insert(id, nama); max_insert(id, nama)
print(f"Data awal berhasil dimuat ({len(data_awal)} item).")

pilihan = 0
while True:
    print("\n========== MENU HEAP ==========")
    print("1. Tambah data (Min-Heap & Max-Heap)")
    print("2. Tampilkan data ascending (Min-Heap)")
    print("3. Tampilkan data descending (Max-Heap)")
    print("4. Hapus data dari Min-Heap")
    print("5. Hapus data dari Max-Heap")
    print("6. Keluar")
    pilihan = int(input("Pilih menu: "))
    if pilihan == 1:
        id = int(input("ID: ")); nama = input("Nama: ")
        min_insert(id, nama); max_insert(id, nama)
        print("Data berhasil ditambahkan ke Min-Heap dan Max-Heap.")
    elif pilihan == 2: min_print()
    elif pilihan == 3: max_print()
    elif pilihan == 4: min_delete()
    elif pilihan == 5: max_delete()
    elif pilihan == 6: print("Program selesai."); break
    else: print("Pilihan tidak valid!")

File: heap_dataset/heap_excel.java
import java.util.*;

public class heap_excel {
    static int[][] minHeap = new int[200][2];
    static String[] minName = new String[200];
    static int minSize = 0;

    static int[][] maxHeap = new int[200][2];
    static String[] maxName = new String[200];
    static int maxSize = 0;

    // ===== MIN HEAP =====
    static void minSwap(int i, int j) {
        int[] ti = minHeap[i]; minHeap[i] = minHeap[j]; minHeap[j] = ti;
        String ts = minName[i]; minName[i] = minName[j]; minName[j] = ts;
    }
    static void minUp(int i) {
        while (i > 0 && minHeap[(i-1)/2][0] > minHeap[i][0]) { minSwap(i, (i-1)/2); i = (i-1)/2; }
    }
    static void minDown(int i) {
        int n = minSize;
        while (true) {
            int s = i, l = 2*i+1, r = 2*i+2;
            if (l < n && minHeap[l][0] < minHeap[s][0]) s = l;
            if (r < n && minHeap[r][0] < minHeap[s][0]) s = r;
            if (s == i) break;
            minSwap(i, s); i = s;
        }
    }
    static void minInsert(int id, String name) {
        minHeap[minSize][0] = id; minName[minSize] = name;
        minUp(minSize++);
    }
    static void minDelete() {
        if (minSize == 0) { System.out.println("Min-Heap kosong!"); return; }
        Scanner sc = new Scanner(System.in);
        System.out.print("Masukkan ID yang ingin dihapus dari Min-Heap: ");
        int id = sc.nextInt();
        int idx = -1;
        for (int i = 0; i < minSize; i++) if (minHeap[i][0] == id) { idx = i; break; }
        if (idx == -1) { System.out.println("ID tidak ditemukan!"); return; }
        System.out.println("Data dihapus: " + minHeap[idx][0] + " - " + minName[idx]);
        minHeap[idx][0] = minHeap[minSize-1][0]; minName[idx] = minName[minSize-1]; minSize--;
        if (idx < minSize) { minUp(idx); minDown(idx); }
    }
    static void minPrint() {
        if (minSize == 0) { System.out.println("Min-Heap kosong!"); return; }
        int[] ids = new int[minSize]; String[] names = new String[minSize];
        for (int i = 0; i < minSize; i++) { ids[i] = minHeap[i][0]; names[i] = minName[i]; }
        // selection sort ascending
        for (int i = 0; i < minSize-1; i++)
            for (int j = i+1; j < minSize; j++)
                if (ids[i] > ids[j]) { int t = ids[i]; ids[i] = ids[j]; ids[j] = t; String ts = names[i]; names[i] = names[j]; names[j] = ts; }
        System.out.println("\n=== DATA MIN-HEAP (Ascending) ===");
        System.out.printf("%-8s %-15s%n", "ID", "Nama");
        System.out.println("-".repeat(25));
        for (int i = 0; i < minSize; i++) System.out.printf("%-8d %-15s%n", ids[i], names[i]);
    }

    // ===== MAX HEAP =====
    static void maxSwap(int i, int j) {
        int[] ti = maxHeap[i]; maxHeap[i] = maxHeap[j]; maxHeap[j] = ti;
        String ts = maxName[i]; maxName[i] = maxName[j]; maxName[j] = ts;
    }
    static void maxUp(int i) {
        while (i > 0 && maxHeap[(i-1)/2][0] < maxHeap[i][0]) { maxSwap(i, (i-1)/2); i = (i-1)/2; }
    }
    static void maxDown(int i) {
        int n = maxSize;
        while (true) {
            int s = i, l = 2*i+1, r = 2*i+2;
            if (l < n && maxHeap[l][0] > maxHeap[s][0]) s = l;
            if (r < n && maxHeap[r][0] > maxHeap[s][0]) s = r;
            if (s == i) break;
            maxSwap(i, s); i = s;
        }
    }
    static void maxInsert(int id, String name) {
        maxHeap[maxSize][0] = id; maxName[maxSize] = name;
        maxUp(maxSize++);
    }
    static void maxDelete() {
        if (maxSize == 0) { System.out.println("Max-Heap kosong!"); return; }
        Scanner sc = new Scanner(System.in);
        System.out.print("Masukkan ID yang ingin dihapus dari Max-Heap: ");
        int id = sc.nextInt();
        int idx = -1;
        for (int i = 0; i < maxSize; i++) if (maxHeap[i][0] == id) { idx = i; break; }
        if (idx == -1) { System.out.println("ID tidak ditemukan!"); return; }
        System.out.println("Data dihapus: " + maxHeap[idx][0] + " - " + maxName[idx]);
        maxHeap[idx][0] = maxHeap[maxSize-1][0]; maxName[idx] = maxName[maxSize-1]; maxSize--;
        if (idx < maxSize) { maxUp(idx); maxDown(idx); }
    }
    static void maxPrint() {
        if (maxSize == 0) { System.out.println("Max-Heap kosong!"); return; }
        int[] ids = new int[maxSize]; String[] names = new String[maxSize];
        for (int i = 0; i < maxSize; i++) { ids[i] = maxHeap[i][0]; names[i] = maxName[i]; }
        // selection sort descending
        for (int i = 0; i < maxSize-1; i++)
            for (int j = i+1; j < maxSize; j++)
                if (ids[i] < ids[j]) { int t = ids[i]; ids[i] = ids[j]; ids[j] = t; String ts = names[i]; names[i] = names[j]; names[j] = ts; }
        System.out.println("\n=== DATA MAX-HEAP (Descending) ===");
        System.out.printf("%-8s %-15s%n", "ID", "Nama");
        System.out.println("-".repeat(25));
        for (int i = 0; i < maxSize; i++) System.out.printf("%-8d %-15s%n", ids[i], names[i]);
    }

    // ===== INIT DATA =====
    static void initData() {
        int[][] data = {
            {5288,0},{5993,0},{8689,0},{8043,0},{8699,0},{2156,0},{4457,0},{8938,0},{2618,0},{9033,0},
            {9971,0},{3874,0},{5914,0},{2398,0},{3725,0},{5210,0},{7363,0},{7631,0},{4513,0},{5656,0},
            {6453,0},{8783,0},{8194,0},{9783,0},{3685,0},{4490,0},{8294,0},{8563,0},{1070,0},{5408,0},
            {8258,0},{9309,0},{1138,0},{2751,0},{3258,0},{6402,0},{7921,0},{9781,0},{3818,0},{5204,0},
            {6119,0},{1928,0},{4207,0},{7255,0},{5309,0},{2897,0},{8028,0},{1660,0},{3248,0},{5641,0},
            {7376,0},{3525,0},{4492,0},{7187,0},{1305,0},{6602,0},{8153,0},{3561,0},{5082,0},{7151,0},
            {7524,0},{9178,0},{9817,0},{4304,0},{6820,0},{9151,0},{3482,0},{3316,0},{5192,0},{7572,0},
            {7660,0},{9224,0},{5083,0},{6362,0},{6465,0},{9888,0},{4159,0},{4969,0},{5097,0},{6271,0},
            {9250,0},{3409,0},{4577,0},{6244,0},{8612,0},{4650,0},{6799,0},{9298,0},{4361,0},{4379,0},
            {6928,0},{3195,0},{5741,0},{6852,0},{8147,0},{8902,0},{8967,0},{1302,0},{2363,0},{6861,0}
        };
        String[] names = {
            "pensil","pulpen","penghapus","buku","sampul","penggaris","kertas","cat","stabilo","mobil",
            "motor","becak","sepeda","kereta","pesawat","perahu","kapal","rakit","kipas","charger",
            "peci","sarung","sajadah","smartphone","jam","televisi","laptop","komputer","mouse","keyboard",
            "tablet","jendela","kaca","pintu","kompor","lemari","kasur","ranjang","bantal","baju",
            "kaos","celana","mukena","jilbab","pigura","antena","kulkas","dispenser","meja","kursi",
            "kemoceng","sapu","gayung","sabun","sikat","shampo","botol","gelas","piring","panci",
            "wajan","blender","galon","cobek","termos","kran","selang","karpet","tikar","keset",
            "sepatu","kaos kaki","jaket","piama","piano","gitar","angklung","suling","toples","parfum",
            "sisir","topi","gunting","pisau","kaleng","tisu","tas","ikat pinggang","korek api","kopi",
            "gula","cabai","wortel","timun","apel","jeruk","tomat","pisang","pepaya","bawang"
        };
        for (int i = 0; i < data.length; i++) { minInsert(data[i][0], names[i]); maxInsert(data[i][0], names[i]); }
        System.out.println("Data awal berhasil dimuat (" + data.length + " item).");
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        initData();
        int pilihan;
        do {
            System.out.println("\n========== MENU HEAP ==========");
            System.out.println("1. Tambah data (Min-Heap & Max-Heap)");
            System.out.println("2. Tampilkan data ascending (Min-Heap)");
            System.out.println("3. Tampilkan data descending (Max-Heap)");
            System.out.println("4. Hapus data dari Min-Heap");
            System.out.println("5. Hapus data dari Max-Heap");
            System.out.println("6. Keluar");
            System.out.print("Pilih menu: ");
            pilihan = sc.nextInt();
            switch (pilihan) {
                case 1:
                    System.out.print("ID: "); int id = sc.nextInt();
                    System.out.print("Nama: "); String nama = sc.next();
                    minInsert(id, nama); maxInsert(id, nama);
                    System.out.println("Data berhasil ditambahkan ke Min-Heap dan Max-Heap.");
                    break;
                case 2: minPrint(); break;
                case 3: maxPrint(); break;
                case 4: minDelete(); break;
                case 5: maxDelete(); break;
                case 6: System.out.println("Program selesai."); break;
                default: System.out.println("Pilihan tidak valid!");
            }
        } while (pilihan != 6);
        sc.close();
    }
}